### PR TITLE
src: Move the statistics option as a suboption under the report

### PR DESCRIPTION
**Pull request by caiquecamargo**
Add a condition while parsing the arguments for the report feature that calls the statistics function when the flag --statistics is
present, as well stopping the user from calling the statistics feature directly.

This is a temporary solution, since both features require modernization on their parsing system.

### CODE REVIEW

**Review by kwy95 — CHANGES_REQUESTED** [review]
As a temporary solution this looks ok, though I am not sure there is a rush to change this behavior until we update the parsers for statistics and report. You also need to update the autocomplete file, and the documentation and help text.

**Comment by rodrigosiqueira** [comment]
Hi @caiquecamargo ,
From our previous discussion, we agreed on this direction:

1. Modernize report parser
2. Move the statistics option under the report

Any update here?

**Comment by caiquecamargo** [comment]
Hi @rodrigosiqueira we've sent an update as a draft so we can discuss better on this and see if we are on the right way with the implementation.

Some tests are break but we working on it.

**Review by kwy95 — CHANGES_REQUESTED** [review]
You removed the statistics option from the parser in the last commit. Also, try to start updating the documentation accordingly.

File: src/report.sh [hunk 1]
```diff
@@ -300,19 +310,21 @@ function save_data_to()
...
-  local week
-  local month
-  local year
   local output
-  local reference=0
+  local reference='DAY'
+  local reference_count=0
+  local options
+  local long_options='day,week,month,year,output,statistics'
+  local short_options='o'
```
Comment (rodrigosiqueira) [inline src/report.sh:320]: The parameter `output | o` requires a value; in this case, you need to add `:` to indicate that.

Additionally, we need to handle values around `day, week, month, year`, but those are optional. For this reason, you can use `::`. However, we will need to use `=`  to set a value.

File: src/report.sh [hunk 2]
```diff
@@ -321,101 +333,100 @@ function report_parse()
...
+  if [[ "$reference_count" -gt 1 ]]; then
     complain 'Please, only use a single time reference'
     return 22
-  elif [[ "$reference" == 0 ]]; then
-    # If no option, set day as a default
+  fi
+
+  if [[ "$output" == 1 ]]; then
+    if [[ "$is_statistics" == 1 ]]; then
+      options_values['ERROR']="--output cannot be used with --statistics."
```
Comment (rodrigosiqueira) [inline src/report.sh:392]: Why we cannot use `--output` with statistics? ihmo, it makes sense.

File: src/report.sh [hunk 3]
```diff
@@ -321,101 +333,100 @@ function report_parse()
...
+
+    if [[ "$#" == 1 ]]; then
+      options_values['OUTPUT']="$1"
+      shift
+    elif [[ "$#" == 2 ]]; then
+      options_values['OUTPUT']="$2"
+    else
+      options_values['ERROR']="Invalid path for the --output argument."
+      return 22 # EINVAL
+    fi
```
Comment (rodrigosiqueira) [inline src/report.sh:404]: What are all of these checks around the output parameter?

File: src/report.sh [hunk 4]
```diff
@@ -321,101 +333,100 @@ function report_parse()
...
+  fi
+
+  if [[ "$#" == 0 && "$reference" == 'DAY' ]]; then
     options_values['DAY']=$(get_today_info '+%Y/%m/%d')
+  elif [[ -n $1 ]]; then
```
Comment (rodrigosiqueira) [inline src/report.sh:409]: Why are you checking `$1`? At this point, I expected that `$@` is empty; if not, something might be wrong.

File: src/statistics.sh [hunk 5]
```diff
@@ -17,83 +17,21 @@ declare -gA shared_data=(['deploy']='' ['build']='' ['list']='' ['uninstall']=''
...
-  if [[ -z "$target" ]]; then
-    target='--day'
+  if [[ -n "${options_values['DAY']}" ]]; then
+    day_statistics "${options_values['DAY']}"
+  elif [[ -n "${options_values['WEEK']}" ]]; then
+    week_statistics "${options_values['WEEK']}"
+  elif [[ -n "${options_values['MONTH']}" ]]; then
+    month_statistics "${options_values['MONTH']}"
+  elif [[ -n "${options_values['YEAR']}" ]]; then
+    year_statistics "${options_values['YEAR']}"
```
Comment (rodrigosiqueira) [inline src/statistics.sh:33]: Currently, `statistics` is the entry point in the `kw` file for the statistic feature. After your refactor, I expect that this code manipulation will abode in the report file.

File: documentation/man/features/report.rst [hunk 6]
```diff
@@ -39,5 +39,59 @@ OPTIONS
 \--year [<year>]:
   Exhibits the current year summary if the user does not specify *<year>*.
 
+\--statistics:
```
Comment (rodrigosiqueira) [inline documentation/man/features/report.rst:42]: This commit should be squashed with the one that enables the `--statistic` option inside the report.

File: documentation/man/kw.rst [hunk 7]
```diff
@@ -100,7 +100,7 @@ the previous sections.
   | :ref:`kw-configm<configm-doc>`
   | :ref:`kw-vars<vars-doc>`
   | :ref:`kw-diff<diff-doc>`
-  | :ref:`kw-statistics<statistics-doc>`
+  | :ref:`kw-report<report-doc>`
```
Comment (rodrigosiqueira) [inline documentation/man/kw.rst:103]: Squash this commit with the one that enables `--statistic` inside `--report`.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Hi,

I added a couple of comments in the PR and follows other requests:

1. You just need two commits in this PR:
 * Commit 1: Modernize report parser
 * Commit 2: Move --statistics inside the report
2. After your PR, I expect the `statistics.sh` file to become some sort of library that can be used for other parts of the code.
3. In the report section, you will need to add some optional values (see my comments).
4. Make sure that you update the test for this new parser option.

File: src/report.sh [hunk 8]
```diff
@@ -321,101 +333,100 @@ function report_parse()
...
+      return 22 # EINVAL
+    fi
+
+    if [[ "$#" == 1 ]]; then
+      options_values['OUTPUT']="$1"
+      shift
+    elif [[ "$#" == 2 ]]; then
+      options_values['OUTPUT']="$2"
+    else
+      options_values['ERROR']="Invalid path for the --output argument."
```
Comment (kwy95) [inline src/report.sh:402]: Single quotes.

**Review by kwy95 — CHANGES_REQUESTED** [review]
@rodrigosiqueira already covered most problems, I left only a few extra comments.

File: src/report.sh [hunk 9]
```diff
@@ -12,15 +12,20 @@ declare -gA options_values
...
   local target_time
   local ret
 
-  report_parse "$@"
-  ret="$?"
-  if [[ "$ret" != 0 ]]; then
-    return "$ret"
+  if [[ "$1" =~ -h|--help ]]; then
+    report_help "$1"
+    return
```
Comment (rodrigosiqueira) [inline src/report.sh:24]: Use `exit` here.

File: src/report.sh [hunk 10]
```diff
@@ -28,7 +30,24 @@ function report_main()
     return 22 # EINVAL
   fi
 
-  if [[ -n "${options_values['DAY']}" ]]; then
+  if [[ "$is_statistics" == 1 ]]; then
```
Comment (rodrigosiqueira) [inline src/report.sh:33]: How about use `options_values['STATISTICS']`?

File: src/report.sh [hunk 11]
```diff
@@ -28,7 +30,24 @@ function report_main()
...
+    if [[ -n "${options_values['DAY']}" ]]; then
+      day_statistics "${options_values['DAY']}"
+    elif [[ -n "${options_values['WEEK']}" ]]; then
+      week_statistics "${options_values['WEEK']}"
+    elif [[ -n "${options_values['MONTH']}" ]]; then
+      month_statistics "${options_values['MONTH']}"
+    elif [[ -n "${options_values['YEAR']}" ]]; then
+      year_statistics "${options_values['YEAR']}"
+    fi
+    return
```
Comment (rodrigosiqueira) [inline src/report.sh:49]: Let's create a function for this code.

File: src/report.sh [hunk 12]
```diff
@@ -308,7 +327,7 @@ function save_data_to()
 function parse_report_options()
 {
   local reference_count=0
-  local long_options='day::,week::,month::,year::,output:'
+  local long_options='day::,week::,month::,year::,output:,statistics'
```
Comment (rodrigosiqueira) [inline src/report.sh:330]: add a short_options to `statistics`, maybe `-s`.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

**Review by kwy95 — CHANGES_REQUESTED** [review]

**Comment by rodrigosiqueira** [comment]
Please, update your PR with the latest code from the unstable branch, addressing Rubens's comments and fixing the formatter-shfmt issue.

**Review by kwy95 — CHANGES_REQUESTED** [review]
This is a nice update. I think it would be interesting as a follow up from this to better integrate the statistics output to the report, so that the statistics info is added to the report, I think this would also solve the conflict of using the `--output` option with statistics.

File: documentation/man/kw.rst [hunk 13]
```diff
@@ -288,20 +288,20 @@ After you start your VM you can ssh into it with::
   kw s -c="dmesg -wH"
   kw s
 
-You can see data related to your kw usage by using the statistics option, see
+You can see data related to your kw usage by using the --statistics flag on the report option, see
```
[the remark this answers is not in the document]
Comment (lucastheo) [reply documentation/man/kw.rst]: We opened issue #523 to solve the issue between output and statistics

File: src/report.sh [hunk 14]
```diff
@@ -386,9 +411,21 @@ function parse_report_options()
         shift 2
         ;;
       --output | -o)
+        if [[ "${options_values['STATISTICS']}" == 1 ]]; then
+          options_values['ERROR']='--output cannot be used with --statistics.'
+          return 22 # EINVAL
+        fi
```
Comment (kwy95) [inline src/report.sh:417]: Check this condition after parsing all arguments, as to not impose an order on the options.

File: src/report.sh [hunk 15]
```diff
@@ -386,9 +411,21 @@ function parse_report_options()
...
         options_values['OUTPUT']="$2"
         shift 2
         ;;
+      --statistics | -s)
+        if [[ -n "${options_values['OUTPUT']}" ]]; then
+          options_values['ERROR']='--output cannot be used with --statistics.'
+          return 22 # EINVAL
+        fi
```
Comment (kwy95) [inline src/report.sh]: Same with this check. Ideally try to unify them.

**Review by kwy95 — CHANGES_REQUESTED** [review]
Small fixes.

File: documentation/man/features/report.rst [hunk 16]
```diff
@@ -19,25 +19,46 @@ processing that will end up as a report for the user.
...
+\--year[=<year>]:
   Exhibits the current year summary if the user does not specify *<year>*.
 
 \--output <file-path>:
   Save the output of the report to *<file-path>*.
+
+EXAMPLES
+========
+For these examples, we suppose the fields in your **kworkflow.config** file are
+already configured.
```
Comment (rodrigosiqueira) [inline documentation/man/features/report.rst:85]: Maybe we can make this comment as a note and rewrite it as well. Maybe something like this:
> First of all, for these examples, we suppose that you have the `disable_statistics_data_track` option in your kworkflow.config enabled for a while.

Comment (rodrigosiqueira) [reply documentation/man/features/report.rst:85]: Also, keep in mind that this is only related to the `--statistics` option. In other words, this is not valid for the regular report; please, make this distinction explicit to the `--statistics` option.

File: src/report.sh [hunk 17]
```diff
@@ -427,9 +417,9 @@ function report_help()
...
   fi
   printf '%s\n' 'kw report:' \
-    '  report [--day [<year>/<month>/<day>]] - Report of the day' \
-    '  report [--week [<year>/<month>/<day>]] - Report of the week' \
-    '  report [--month [<year>/<month>]] - Report of the month' \
-    '  report [--year [<year>]] - Report fo the year' \
+    '  report [--day[=<year>/<month>/<day>]] - Report of the day' \
+    '  report [--week[=<year>/<month>/<day>]] - Report of the week' \
+    '  report [--month[=<year>/<month>]] - Report of the month' \
+    '  report [--year[=<year>]] - Report fo the year' \
```
Comment (rodrigosiqueira) [inline src/report.sh:457]: Drop the `statistics` option in the main help menu. See `src/help.sh` file.

File: documentation/man/features/report.rst [hunk 18]
```diff
@@ -39,6 +53,29 @@ OPTIONS
...
+  Shows the week summary. If a user does not pass any parameter kw will show
+  the current week statistics. However, users can pass a random date
+  *<year>/<month>/<day>* and let **kw** take care to provide a summary
+  related to the week related to the target date.
+
+\--statistics [--month[=<year>/<month>]]:
+  Shis option shows a report regarding a specific month. Users can search for
+  data related to a specific month by providing a parameter in the
+  *<year>/<month>* format. If the user does not pass any parameter, **kw**
+  displays the current month data.
```
Comment (rodrigosiqueira) [inline documentation/man/features/report.rst:74]: The option `kw report --help` will looks like this for this section:
```
--day[=<year>/<month>/<day>]:
Display day summary. Users have the option to search a specific date by passing an argument that follows the <year>/<month>/<day> format. Not passing anything and gets info about today.

--week[=<year>/<month>/<day>]:
Shows the week summary. If a user does not pass any parameter kw will show the current week statistics. However, users can pass a random date <year>/<month>/<day> and let kw take care to provide a summary related to the week related to the target date.

--month[=<year>/<month>]:
This option shows a report regarding a specific month. Users can search for data related to a specific month by providing a parameter in the <year>/<month> format. If the user does not pass any parameter, kw displays the current month data.

--year[=<year>]:
Exhibits the current year summary if the user does not specify <year>.

--statistics:
Display current date statistics summary.

--statistics [–day[=<year>/<month>/<day>]]:
Display day statistics summary. Users have the option to search a specific date by passing an argument that follows the <year>/<month>/<day> format. Not passing anything and gets info about today.

--statistics [–week[=<year>/<month>/<day>]]:
Shows the week summary. If a user does not pass any parameter kw will show the current week statistics. However, users can pass a random date <year>/<month>/<day> and let kw take care to provide a summary related to the week related to the target date.

--statistics [–month[=<year>/<month>]]:
Shis option shows a report regarding a specific month. Users can search for data related to a specific month by providing a parameter in the <year>/<month> format. If the user does not pass any parameter, kw displays the current month data.

--statistics [–year[=<year>]]:
Exhibits the current year summary if the user does not specify <year>.
```
I don't think it makes sense to explain the date parameters repeatedly. Please, rewrite it by removing the duplication of information.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

**Comment by rodrigosiqueira** [comment]
@caiquecamargo any update here?

**Comment by caiquecamargo** [comment]
> @caiquecamargo any update here?

We resolve the requested changes.

File: documentation/man/features/report.rst [hunk 19]
```diff
@@ -39,13 +55,15 @@ OPTIONS
 \--year[=<year>]:
   Exhibits the current year summary if the user does not specify *<year>*.
 
+\--statistics:
+  Display current date statistics summary.
```
Comment (kwy95) [inline documentation/man/features/report.rst:59]: Change this to be more generic and take into account the date parameters. Something like this:
`Display statistics for the given date.`
You could alternatively just add something like `You can choose the date and range with the date options.`

**Review by kwy95 — CHANGES_REQUESTED** [review]
Minor fixes.

**Comment by rodrigosiqueira** [comment]
I changed some details and merged this PR with the unstable branch.
Thanks

**Pull request closed without merging**